Providers/RustSyncManager.swift

```diff
@@ -0,0 +1,628 @@
+                       level: .debug,
+                       category: .sync)
+            return
+        }
+
+        let interval = fifteenMinutesInterval
+        let selector = #selector(syncOnTimer)
+        logger.log("Starting sync timer.",
+                   level: .info,
+                   category: .sync)
+        self.syncTimer = repeatingTimerAtInterval(interval, selector: selector)
+    }
+
+    /**
+     * The caller is responsible for calling this on the same thread on which it called
+     * beginTimedSyncs.
+     */
+    public func endTimedSyncs() {
+        if let t = self.syncTimer {
+            logger.log("Stopping sync timer.",
+                       level: .info,
+                       category: .sync)
+            self.syncTimer = nil
+            t.invalidate()
+        }
+    }
+
+    public func applicationDidBecomeActive() {
+        backgrounded = false
+
+        guard self.profile.hasSyncableAccount() else { return }
+
+        self.beginTimedSyncs()
+
+        // Sync now if it's been more than our threshold.
+        let now = Date.now()
+        let then = self.lastSyncFinishTime ?? 0
+        guard now >= then else {
+            logger.log("Time was modified since last sync.",
+                       level: .debug,
+                       category: .sync)
+            self.syncEverythingSoon()
+            return
+        }
+        let since = now - then
+        logger.log("\(since)msec since last sync.",
+                   level: .debug,
+                   category: .sync)
+        if since > SyncConstants.SyncOnForegroundMinimumDelayMillis {
+            self.syncEverythingSoon()
+        }
+    }
+
+    public func applicationDidEnterBackground() {
+        backgrounded = true
+    }
+
+    fileprivate func beginSyncing() {
+        notifySyncing(notification: .ProfileDidStartSyncing)
+    }
+
+    private func resolveSyncState(
+        result: MozillaAppServices.SyncResult
+    ) -> SyncDisplayState {
+        let hasSynced = !result.successful.isEmpty
+        let status = result.status
+
+        // This is similar to the old `SyncStatusResolver.resolveResults` call. If none of
+        // the engines successfully synced and a network issue occured we return `.bad`.
+        // If none of the engines successfully synced and an auth error occured we return
+        // `.warning`. Otherwise we return `.good`.
+
+        if !hasSynced && status == .authError {
+            return .warning(message: .FirefoxSyncOfflineTitle)
+        } else if !hasSynced && status == .networkError {
+            return .bad(message: .FirefoxSyncOfflineTitle)
+        } else {
+            return .good
+        }
+    }
+
+    fileprivate func endRustSyncing(_ result: MozillaAppServices.SyncResult) {
+        logger.log("Ending all syncs.",
+                   level: .info,
+                   category: .sync)
+
+        syncDisplayState = resolveSyncState(result: result)
+
+        #if MOZ_TARGET_CLIENT
```

Comment:
Do we need this and the `canSendUsageData`? I thought that with glean, telemetry settings are global (i.e glean won't send anything at all if disabled)?

Reply:
Bet! I'll remove this.